doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,245 @@
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+__owur int ossl_quic_packetise(SSL *stream, const OSSL_QUIC_FRAME *frame,
+                               size_t frame_length);
+```
+
+#### Notes
+
+- Should more than one frame be queueable at once?
+- More arguments are likely going to be required.
+- How to destroy frames?  `OPENSSL_free` or a callback?
+
+### TLS Handshake Record Layer
+
+Uses the Record Layer API to implement the inner TLS-1.3 protocol handshake.
+It produces the QUIC crypto frames which are queued using the same mechanism
+as the [Stream Send Buffers](#stream-send-buffers) above.
+
+#### Notes
+
+- How distinguish handshake from rest?
+
+### Flow Controller and Statistics Collector
+
+To make decisions about what frames to coalesce, the packetiser relies
+on the flow controller to enforce stream and connection bandwidth limits
+[RFC 9000 4.1 Data Flow Control].
+
+```c
+/* API to be defined for flow controller */
+```
+
+### Congestion Controller
+
+Also part of the frame coalescing decision is the congestion controller
+[RFC 9002].  For MVP, this will be a _just send it_.
+
+```c
+/* Pluggable congestion controller APIs go here */
+```
+
+### QUIC Write Record Layer
+
+Coalesced frames are passed to the QUIC record layer for encryption and sending.
+To send accumulated frames as packets to the QUIC Write Record Layer:
+
+```c
+__owur int ossl_quic_send_packets(SSL *connection);
+```
+
+#### Notes
+
+- Should the send call accept a stream SSL, a connection SSL or either?
+- How to handle multiple security levels in a packet?
+- Who should call the send function?  A short timer to allow more data to be accumulated in the packetise function?
+
+### ACK Handline and Loss Detector
+
+1. When a packet is sent, the packetiser needs to inform the ACK Manager.
+2. When a packet is ACKed, the packetiser is informed so it can drop sent frames.
+3. When a packet is lost, the packetiser is informed so it can create new packet(s) for retransmissions.
+4. When a packet is discarded without ACK or loss, the packetiser is informed so it can clean up.
+
+```c
+__owur int ossl_quic_packet_sent(SSL *connection, OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_acked(SSL *connection, OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_lost(SSL *connection, OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_discarded(SSL *connection, OSSL_QUIC_PACKET *packet);
+```
+
+#### Notes
+
+| Name here | Name in ACK Manager |
+| --- | --- |
+| `ossl_quic_packet_sent` | `QUIC_ACKM_on_tx_ack_packet` |
+| `ossl_quic_packet_acked` | `on_acked` |
+| `ossl_quic_packet_lost` | `on_lost` |
+| `ossl_quic_packet_discarded` | `on_discarded` |
+
+- Do the latter three need to be function pointers?
+- What is the benefit of having them as such?
```

Comment:
It keeps the ACK manager decoupled from other modules. This is especially helpful given that we're developing these modules independently, so there are issues with having them depend on one another directly.

I guess once things are more tied down it can be hardcoded. There are examples of both approaches in the wild (hardcoded: MSQUIC, callback: go-quic).